COMPAT: unique() should preserve the dtype of the input

stuarteberg opened the pull request:
The behavior of `pd.unique()` is surprising, because -- unlike `np.unique()` -- the result does not have the same `dtype` as the input:

```python
In [1]: pd.Series([1,2,3], dtype=np.uint8).unique()
Out[1]: array([1, 2, 3], dtype=uint64)
```

This PR just casts the output array to match the input dtype.  Supercedes #27869.

**Update:** Augmented the tests to cover narrow dtypes.

~I added a new assertion in `test_value_counts_unique_nunique()`, but it may not be sufficient.  From what I can see, there isn't good coverage of `Series` whose *data* is not `int`/`float`/ etc.  There is only good coverage of various *index* types. Any advice concerning test coverage?~

- [x] closes #27869
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

stuarteberg committed:
API: unique() should preserve the dtype of the input

jreback reviewed (changes requested):
this would need actual tests for non 64bit dtypes

stuarteberg commented:
>this would need actual tests for non 64bit dtypes

Sounds good.  Is `pandas/tests/test_base.py` the right place for that?  I thought that [`TestIndexOps.test_value_counts_unique_nunique`][1] might be the natural place, but that class's [test members][2] are paramterized by *index* dtype, not *data* dtype.  Is it okay to add a new test member(s) to that class which differs by data dtype, rather than index dtype?  Would that be out-of-place?

[1]: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/test_base.py#L345
[2]: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/test_base.py#L178-L185

stuarteberg commented:
>this would need actual tests for non 64bit dtypes

Okay, I've augmented `test_base` with extra test Series, which use narrow dtypes.

The tests work: Without the new `unique()` behavior, the tests fail.  But with the new change, they pass.

stuarteberg commented:
@jreback I added the requested test cases and rebased.  The CI is passing now.

BTW, I had a problem with an unrelated test.  It was [expected to fail][1], but in this PR it magically started passing -- under Python 3.5 only.  Probably somehow related to dict ordering.  Anyway, I "fixed" the issue by [simply permitting the test to `xpass` under Python 3.5.][2]

FWIW, I'm not the only one who ran into that `xfail` problem. It was also encountered recently in PR #27762.

[1]: https://github.com/pandas-dev/pandas/blob/5de4e55d60bf8487a2ce64a440b6d5d92345a4bc/pandas/tests/series/test_analytics.py#L1492

[2]: https://github.com/pandas-dev/pandas/pull/27874/files#diff-dc347bc3d0448ea297bed67dc7ff3437R1494

TomAugspurger commented on pandas/tests/series/test_analytics.py:
```diff
@@ -1489,7 +1490,8 @@ def test_value_counts_with_nan(self):
             pytest.param(
-                "datetime64[D]", marks=pytest.mark.xfail(reason="GH#7996", strict=True)
+                "datetime64[D]",
+                marks=pytest.mark.xfail(reason="GH#7996", strict=not PY35),
```
I don't understand this xfail. Typically we just reference open issues. What's causing this to fail on 3.5? What's the failure you get?

stuarteberg replied:
>What's causing this to fail on 3.5?

Just to be clear: The problem is that this test *doesn't* fail in Python 3.5.  But since it's marked with `xfail(..., strict=True)`, that breaks the test suite.

You can see a failed build log here:
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=15986

The relevant lines are:
```
=================================== FAILURES ===================================
 TestCategoricalSeriesAnalytics.test_drop_duplicates_categorical_non_bool[True-datetime64[D]] 
[gw1] linux -- Python 3.5.3 /home/vsts/miniconda3/envs/pandas-dev/bin/python
[XPASS(strict)] GH#7996
 TestCategoricalSeriesAnalytics.test_drop_duplicates_categorical_non_bool[False-datetime64[D]] 
[gw1] linux -- Python 3.5.3 /home/vsts/miniconda3/envs/pandas-dev/bin/python
[XPASS(strict)] GH#7996
 TestCategoricalSeriesAnalytics.test_drop_duplicates_categorical_non_bool[None-datetime64[D]] 
[gw1] linux -- Python 3.5.3 /home/vsts/miniconda3/envs/pandas-dev/bin/python
[XPASS(strict)] GH#7996
-------- generated xml file: /home/vsts/work/1/s/test-data-multiple.xml --------
```

>What's the failure you get?

In Python 3.7, the test `xfails` as expected.  Removing `xfail` identifies [this line as the problem][1].  And here's what `pytest` shows in that case:

```
>   raise_assert_detail(obj, msg, lobj, robj)
E   AssertionError: Series are different
E
E   Series values are different (50.0 %)
E   [left]:  [False, True, True, True]
E   [right]: [False, False, False, True]

pandas/_libs/testing.pyx:178: AssertionError
```

[1]: https://github.com/pandas-dev/pandas/blob/11b28ea584b3647bc8ca5b56889c9199c0af5ec5/pandas/tests/series/test_analytics.py#L1504

stuarteberg replied:
>I don't understand this xfail. Typically we just reference open issues.

The referenced issue seems to imply that the trouble is related to converting/comparing  `datetime[D]` to `datetime[ns]`.  In this test, the input is `datetime[D]`, but it's implicitly converted to `datetime[ns]` when it is loaded into a `Categorical`.

One simple hack to make this test pass is to change `datetime[D]` to `datetime[ns]`.  That doesn't seem appropriate, though.

Anyway, to get an idea of what is actually going wrong, here's what happens when I try the test's first three lines in my terminal.  Note that the first item becomes `NaT` for some reason.

```
In [190]: cat_array = np.array([1, 2, 3, 4, 5], dtype=np.dtype(dtype))

In [191]: input1 = np.array([1, 2, 3, 3], dtype=np.dtype(dtype))

In [192]: tc1 = Series(Categorical(input1, categories=cat_array, ordered=False))

In [193]: tc1
Out[193]:
0          NaT
1   1970-01-03
2   1970-01-04
3   1970-01-04
dtype: category
Categories (5, datetime64[ns]): [1970-01-02, 1970-01-03, 1970-01-04, 1970-01-05, 1970-01-06]
```

TomAugspurger commented on doc/source/whatsnew/v0.25.1.rst:
```diff
@@ -156,7 +156,7 @@ ExtensionArray
 ^^^^^
 - Bug in :meth:`Series.replace` and :meth:`DataFrame.replace` when replacing timezone-aware timestamps using a dict-like replacer (:issue:`27720`)
--
+- The returned dtype of ::func:`pd.unique` now matches the input dtype. (:issue`#27874`)
```
@jreback thoughts on 0.25.1 or 1.0 for this? I'm leaning towards 1.0

TomAugspurger replied:
And @stuarteberg you can remove the `#` from the issue number.

jreback commented on pandas/core/algorithms.py:
```diff
@@ -404,7 +404,7 @@ def unique(values):
     table = htable(len(values))
     uniques = table.unique(values)
-    uniques = _reconstruct_data(uniques, dtype, original)
+    uniques = _reconstruct_data(uniques, original.dtype, original)
```
in _reconstruct_data can you change this
```
values = values.astype(dtype)
```
to pass ```copy=False```

jreback commented on pandas/tests/test_base.py:
```diff
@@ -187,7 +187,24 @@ def setup_method(self, method):
         self.series = [getattr(self, "{}_series".format(t)) for t in types]
-        self.objs = self.indexes + self.series
+
+        # To test narrow dtypes, we use narrower *data* elements, not *index* elements
```
this whole thing badly needs parameterization.

can you pull out the unique tests and do that instead of repeating all of this?

jreback commented:
can you merge master; update the release note to 1.0, will have a look after

stuarteberg committed:
_reconstruct_data(): Use copy=False when calling astype()

stuarteberg commented:
Sorry I haven't had time to look at this.  I'll try to get to it next week.

jreback commented:
@stuarteberg looked reasonable, can you merge master and move the note to 1.0.0

stuarteberg commented:
@jreback 

>looked reasonable, can you merge master and move the note to 1.0.0

Done.  Sorry I haven't cleaned up the tests yet.  If its sufficient as-is, great.  If not, let me know.

jreback commented:
thanks @stuarteberg this looks great! Can you create an issue to parameterize test_base (where indicated)? A PR would be most appreciated for that as well (if you can).

jreback commented:
this did not have any issue (other than the PR itself) associated? can you do a quick search to see if this solves any open issues?

stuarteberg commented:
>this did not have any issue (other than the PR itself) associated?
>can you do a quick search to see if this solves any open issues?

OK, I couldn't find an issue for this in particular, but I just found #22824, which has a much broader scope.  It does mention the return type issue:

>- [ ] Change return type for [Series/Index].unique to be same as caller (deprecation cycle by introducing raw=None which at first defaults to True?)

But since this PR doesn't resolve everything mentioned in that issue, it should remain open.  But I made a comment there referencing this PR.

>Can you create an issue to parameterize test_base (where indicated)?

OK, I have an issue draft ready to go (pasted into the details section below), but github noticed that #23877 is similar.  Is that already sufficient, or would you like me to open the new issue anyway?

<details>

<summary>Issue draft regarding test_base</summary>

The tests in `tests/test_base.py` exercise the behavior of both Series and Indexes of various dtypes, but without using the standard `pytest` mechanisms for parameterization.  In particular, the [setup of the `Ops` base class][1] should be cleaned up (or removed) in favor of proper pytest fixtures.

[1]: https://github.com/pandas-dev/pandas/blob/a1dba2c6d6140896cdb23d9c9f094d8c71094ccf/pandas/tests/test_base.py#L167

FWIW, this was [noticed][2] while reviewing #27874, so @jreback [requested][3] this issue to be opened.

[2]: https://github.com/pandas-dev/pandas/pull/27874#discussion_r314311204
[3]: https://github.com/pandas-dev/pandas/pull/27874#issuecomment-538802943

</details>

jreback commented:
@stuarteberg no #23877 is ok

jreback commented:
thanks @stuarteberg
